File: webapp-vercel/lib/analysts.ts
import { sma, rsi, momentum, sampleStdev } from './indicators';
import type { AnalystReport, MarketSnapshot } from './types';

function scoreToSignal(score: number, maxAbsScore: number) {
  const confidence = maxAbsScore ? Math.min(1.0, Math.abs(score) / maxAbsScore) : 0.0;
  if (score > 0.25) return { signal: 'bullish' as const, confidence };
  if (score < -0.25) return { signal: 'bearish' as const, confidence };
  return { signal: 'neutral' as const, confidence };
}

export function technicalAnalyst(snapshot: MarketSnapshot): AnalystReport {
  const closes = snapshot.closes;
  const fast = sma(closes, 10), slow = sma(closes, 30), rsiV = rsi(closes, 14), mom = momentum(closes, 10);
  const points: string[] = [];
  let score = 0;
  if (fast !== null && slow !== null) {
    if (fast > slow) { score += 1; points.push(`SMA10 (${fast.toFixed(2)}) above SMA30 (${slow.toFixed(2)}): uptrend`); }
    else { score -= 1; points.push(`SMA10 (${fast.toFixed(2)}) below SMA30 (${slow.toFixed(2)}): downtrend`); }
  }
  if (rsiV !== null) {
    if (rsiV >= 70) { score -= 0.5; points.push(`RSI ${rsiV.toFixed(1)}: overbought, pullback risk`); }
    else if (rsiV <= 30) { score += 0.5; points.push(`RSI ${rsiV.toFixed(1)}: oversold, bounce potential`); }
    else points.push(`RSI ${rsiV.toFixed(1)}: neutral`);
  }
  if (mom !== null) { score += mom > 0 ? 0.5 : -0.5; points.push(`10-bar momentum ${(mom * 100).toFixed(1)}%`); }
  const { signal, confidence } = scoreToSignal(score, 2.0);
  return { agent_name: 'technical_analyst', signal, confidence, summary: points[0] || 'No indicator data available.', key_points: points };
}

export function fundamentalAnalyst(snapshot: MarketSnapshot): AnalystReport {
  const f = snapshot.fundamentals;
  const points: string[] = [];
  let score = 0;
  const pe = f.pe_ratio;
  if (pe !== undefined) {
    if (pe < 15) { score += 1; points.push(`P/E ${pe.toFixed(1)}: reasonably valued or cheap`); }
    else if (pe > 30) { score -= 1; points.push(`P/E ${pe.toFixed(1)}: richly valued`); }
    else points.push(`P/E ${pe.toFixed(1)}: fair value`);
  }
  const growth = f.revenue_growth_yoy;
  if (growth !== undefined) {
    score += growth > 0.10 ? 1 : growth < 0 ? -0.5 : 0;
    points.push(`YoY revenue growth ${(growth * 100).toFixed(1)}%`);
  }
  const fpe = f.forward_pe;
  if (fpe !== undefined && pe !== undefined && pe > 0) {
    if (fpe < pe) { score += 0.5; points.push(`Forward P/E ${fpe.toFixed(1)} < trailing ${pe.toFixed(1)}: earnings expected to grow`); }
    else if (fpe > pe) { score -= 0.5; points.push(`Forward P/E ${fpe.toFixed(1)} > trailing ${pe.toFixed(1)}: earnings expected to shrink`); }
  }
  const roe = f.return_on_equity;
  if (roe !== undefined) { score += roe > 0.15 ? 0.5 : roe < 0.05 ? -0.5 : 0; points.push(`Return on equity ${(roe * 100).toFixed(1)}%`); }
  const margin = f.profit_margin;
  if (margin !== undefined) { score += margin > 0.15 ? 0.5 : margin < 0 ? -1 : 0; points.push(`Profit margin ${(margin * 100).toFixed(1)}%`); }
  const dte = f.debt_to_equity;
  if (dte !== undefined) { score += dte > 200 ? -0.5 : dte < 50 ? 0.5 : 0; points.push(`Debt/Equity ${dte.toFixed(0)}%`); }
  const { signal, confidence } = scoreToSignal(score, 5.0);
  return { agent_name: 'fundamental_analyst', signal, confidence, summary: points[0] || 'No fundamental data available.', key_points: points };
}

const POSITIVE_WORDS = ['buy', 'upgrade', 'beat', 'rally', 'surge', 'gain', 'record', 'growth', 'strong', 'outperform', 'raise'];
const NEGATIVE_WORDS = ['sell', 'downgrade', 'miss', 'plunge', 'slump', 'loss', 'warning', 'lawsuit', 'weak', 'underperform', 'cut', 'fall'];

export function sentimentAnalyst(snapshot: MarketSnapshot): AnalystReport {
  const headlines = snapshot.newsHeadlines;
  let score = 0;
  const points: string[] = [];
  for (const h of headlines) {
    const low = h.toLowerCase();
    const pos = POSITIVE_WORDS.some((w) => low.includes(w));
    const neg = NEGATIVE_WORDS.some((w) => low.includes(w));
    if (pos && !neg) { score += 1; points.push(`+ ${h}`); }
    else if (neg && !pos) { score -= 1; points.push(`- ${h}`); }
    else points.push(`= ${h}`);
  }
  const { signal, confidence } = scoreToSignal(score, Math.max(1.0, headlines.length));
  return { agent_name: 'sentiment_analyst', signal, confidence, summary: headlines[0] || 'No headlines available.', key_points: points };
}

export function macroAnalyst(snapshot: MarketSnapshot): AnalystReport {
  const m = snapshot.macro;
  const points: string[] = [];
  let score = 0;
  if (m.ten_year_yield_change_pct !== null) {
    if (m.ten_year_yield_change_pct > 0.05) { score -= 0.5; points.push(`10Y yield up ${(m.ten_year_yield_change_pct * 100).toFixed(1)}%: tightening headwind`); }
    else if (m.ten_year_yield_change_pct < -0.05) { score += 0.5; points.push(`10Y yield down ${(m.ten_year_yield_change_pct * 100).toFixed(1)}%: easier conditions`); }
  }
  if (m.vix_level !== null) {
    if (m.vix_level > 25) { score -= 1; points.push(`VIX ${m.vix_level.toFixed(1)}: elevated fear, risk-off`); }
    else if (m.vix_level < 15) { score += 0.5; points.push(`VIX ${m.vix_level.toFixed(1)}: calm, risk-on`); }
    else points.push(`VIX ${m.vix_level.toFixed(1)}: normal range`);
  }
  if (m.dollar_index_change_pct !== null) {
    if (m.dollar_index_change_pct > 0.02) { score -= 0.5; points.push(`Dollar index up ${(m.dollar_index_change_pct * 100).toFixed(1)}%: headwind`); }
    else if (m.dollar_index_change_pct < -0.02) { score += 0.5; points.push(`Dollar index down ${(m.dollar_index_change_pct * 100).toFixed(1)}%: tailwind`); }
  }
  const { signal, confidence } = scoreToSignal(score, 2.0);
  return { agent_name: 'macro_analyst', signal, confidence, summary: points[0] || 'No macro data available.', key_points: points };
}

const FULL_CONFIDENCE_MOVE = 0.05;
const NEUTRAL_BAND = 0.2 * FULL_CONFIDENCE_MOVE;

export function forecastAnalyst(snapshot: MarketSnapshot): AnalystReport {
  const closes = snapshot.closes;
  const predLen = 10, lookback = 20;
  let expectedReturn = 0, dispersion = 0;
  if (closes.length >= 2) {
    const window = closes.slice(-lookback);
    const returns: number[] = [];
    for (let i = 1; i < window.length; i++) returns.push(window[i] / window[i - 1] - 1);
    const drift = returns.reduce((a, b) => a + b, 0) / (returns.length || 1);
    const vol = sampleStdev(returns);
    let price = closes[closes.length - 1];
    for (let i = 0; i < predLen; i++) price *= 1 + drift;
    expectedReturn = price / closes[closes.length - 1] - 1;
    dispersion = vol * Math.sqrt(predLen);
  }
  const rawConfidence = Math.min(1.0, Math.abs(expectedReturn) / FULL_CONFIDENCE_MOVE);
  let signal: 'bullish' | 'bearish' | 'neutral' = 'neutral';
  if (expectedReturn > NEUTRAL_BAND) signal = 'bullish';
  else if (expectedReturn < -NEUTRAL_BAND) signal = 'bearish';
  const uncertaintyPenalty = 1.0 / (1.0 + dispersion / FULL_CONFIDENCE_MOVE);
  const confidence = rawConfidence * uncertaintyPenalty;
  const summary = `heuristic ${predLen}-bar forecast: ${(expectedReturn * 100).toFixed(1)}% expected move (dispersion ${(dispersion * 100).toFixed(1)}%)`;
  return { agent_name: 'forecast_analyst', signal, confidence, summary, key_points: [summary] };
}

export function runAllAnalysts(snapshot: MarketSnapshot): AnalystReport[] {
  return [technicalAnalyst(snapshot), fundamentalAnalyst(snapshot), sentimentAnalyst(snapshot), macroAnalyst(snapshot), forecastAnalyst(snapshot)];
}

File: trading_agent/agents/analysts.py
"""Analyst agents: turn raw market data into an AnalystReport each.

Each analyst computes its signal with a deterministic rule first (so the
system works and is testable with no LLM at all), then optionally asks
the LLM client for a short human-readable narrative on top. The LLM never
decides the signal/confidence numbers themselves.
"""
from __future__ import annotations

from trading_agent.agents.schemas import AnalystReport, Signal
from trading_agent.data.indicators import momentum, rsi, sma
from trading_agent.data.providers import MarketSnapshot
from trading_agent.forecast.base import PriceForecaster
from trading_agent.llm.client import LLMClient


class TechnicalAnalyst:
    name = "technical_analyst"

    def __init__(self, llm: LLMClient) -> None:
        self._llm = llm

    def analyze(self, snapshot: MarketSnapshot) -> AnalystReport:
        closes = snapshot.closes
        fast = sma(closes, 10)
        slow = sma(closes, 30)
        rsi_value = rsi(closes, 14)
        mom = momentum(closes, 10)

        points: list[str] = []
        score = 0.0
        if fast is not None and slow is not None:
            if fast > slow:
                score += 1
                points.append(f"SMA10 ({fast:.2f}) above SMA30 ({slow:.2f}): uptrend")
            else:
                score -= 1
                points.append(f"SMA10 ({fast:.2f}) below SMA30 ({slow:.2f}): downtrend")
        if rsi_value is not None:
            if rsi_value >= 70:
                score -= 0.5
                points.append(f"RSI {rsi_value:.1f}: overbought, pullback risk")
            elif rsi_value <= 30:
                score += 0.5
                points.append(f"RSI {rsi_value:.1f}: oversold, bounce potential")
            else:
                points.append(f"RSI {rsi_value:.1f}: neutral")
        if mom is not None:
            score += 0.5 if mom > 0 else -0.5
            points.append(f"10-bar momentum {mom * 100:.1f}%")

        signal, confidence = _score_to_signal(score, max_abs_score=2.0)
        summary = self._llm.narrate(
            system="You are a terse technical analyst. One sentence, no advice.",
            user="\n".join(points) or "No indicator data available.",
        )
        return AnalystReport(self.name, signal, confidence, summary, points)


class FundamentalAnalyst:
    name = "fundamental_analyst"

    def __init__(self, llm: LLMClient) -> None:
        self._llm = llm

    def analyze(self, snapshot: MarketSnapshot) -> AnalystReport:
        pe = snapshot.fundamentals.get("pe_ratio")
        growth = snapshot.fundamentals.get("revenue_growth_yoy")
        points: list[str] = []
        score = 0.0
        if pe is not None:
            if pe < 15:
                score += 1
                points.append(f"P/E {pe:.1f}: reasonably valued or cheap")
            elif pe > 30:
                score -= 1
                points.append(f"P/E {pe:.1f}: richly valued")
            else:
                points.append(f"P/E {pe:.1f}: fair value")
        if growth is not None:
            score += 1 if growth > 0.10 else (-0.5 if growth < 0 else 0)
            points.append(f"YoY revenue growth {growth * 100:.1f}%")

        signal, confidence = _score_to_signal(score, max_abs_score=2.0)
        summary = self._llm.narrate(
            system="You are a terse fundamental analyst. One sentence, no advice.",
            user="\n".join(points) or "No fundamental data available.",
        )
        return AnalystReport(self.name, signal, confidence, summary, points)


class SentimentAnalyst:
    name = "sentiment_analyst"

    _POSITIVE_WORDS = ("buying", "return", "rally", "beat", "upgrade", "resistance broken")
    _NEGATIVE_WORDS = ("selloff", "downgrade", "miss", "resistance", "warning", "lawsuit")

    def __init__(self, llm: LLMClient) -> None:
        self._llm = llm

    def analyze(self, snapshot: MarketSnapshot) -> AnalystReport:
        headlines = snapshot.news_headlines
        score = 0.0
        points: list[str] = []
        for headline in headlines:
            lowered = headline.lower()
            pos = any(w in lowered for w in self._POSITIVE_WORDS)
            neg = any(w in lowered for w in self._NEGATIVE_WORDS)
            if pos and not neg:
                score += 1
                points.append(f"+ {headline}")
            elif neg and not pos:
                score -= 1
                points.append(f"- {headline}")
            else:
                points.append(f"= {headline}")

        signal, confidence = _score_to_signal(score, max_abs_score=max(1.0, len(headlines)))
        summary = self._llm.narrate(
            system="You are a terse news-sentiment analyst. One sentence, no advice.",
            user="\n".join(headlines) or "No headlines available.",
        )
        return AnalystReport(self.name, signal, confidence, summary, points)


class ForecastAnalyst:
    """Wraps a `PriceForecaster` (Kronos, or the offline heuristic fallback).

    Like the other analysts, the forecaster only supplies numbers
    (predicted return, sample dispersion); this class turns those into a
    signal/confidence using a fixed, inspectable rule, and the forecaster
    itself never sees or influences position sizing or leverage.
    """

    name = "forecast_analyst"

    # A pred_len-horizon move of this size or larger maps to full-confidence
    # bullish/bearish; smaller moves scale down linearly. Kept conservative
    # since Kronos's own docs are explicit that it forecasts prices, not
    # profitable trades.
    _FULL_CONFIDENCE_MOVE = 0.05
    # Below this fraction of the full-confidence move, treat the forecast
    # as noise rather than a directional call (a return-fraction score
    # needs its own threshold — it isn't comparable to the raw point
    # totals `_score_to_signal`'s fixed 0.25 cutoff was built for).
    _NEUTRAL_BAND = 0.20 * _FULL_CONFIDENCE_MOVE

    def __init__(self, llm: LLMClient, forecaster: PriceForecaster, pred_len: int = 10) -> None:
        self._llm = llm
        self._forecaster = forecaster
        self._pred_len = pred_len

    def analyze(self, snapshot: MarketSnapshot) -> AnalystReport:
        result = self._forecaster.forecast(snapshot.closes, self._pred_len)

        raw_confidence = min(1.0, abs(result.expected_return) / self._FULL_CONFIDENCE_MOVE)
        if result.expected_return > self._NEUTRAL_BAND:
            signal = Signal.BULLISH
        elif result.expected_return < -self._NEUTRAL_BAND:
            signal = Signal.BEARISH
        else:
            signal = Signal.NEUTRAL
        # Wide sample dispersion (the model's own samples disagree on
        # direction/magnitude) should pull confidence toward zero rather
        # than letting a noisy point estimate look decisive.
        uncertainty_penalty = 1.0 / (1.0 + result.dispersion / self._FULL_CONFIDENCE_MOVE)
        confidence = raw_confidence * uncertainty_penalty

        points = [
            f"{result.source} {self._pred_len}-bar forecast: "
            f"{result.expected_return * 100:+.1f}% expected move "
            f"(dispersion {result.dispersion * 100:.1f}%, n={result.sample_count})"
        ]
        summary = self._llm.narrate(
            system="You are a terse quant summarizing a price forecast. One sentence, no advice.",
            user=points[0],
        )
        return AnalystReport(self.name, signal, confidence, summary, points)


def _score_to_signal(score: float, max_abs_score: float) -> tuple[Signal, float]:
    confidence = min(1.0, abs(score) / max_abs_score) if max_abs_score else 0.0
    if score > 0.25:
        return Signal.BULLISH, confidence
    if score < -0.25:
        return Signal.BEARISH, confidence
    return Signal.NEUTRAL, confidence
